plugins/outputs/otlp/otlp.go

```diff
@@ -0,0 +1,379 @@
+)
+
+// OTLP is the OpenTelemetry Protocol config info.
+type OTLP struct {
+	Endpoint   string            `toml:"endpoint"`
+	Timeout    string            `toml:"timeout"`
+	Headers    map[string]string `toml:"headers"`
+	Attributes map[string]string `toml:"attributes"`
+
+	Namespace string
+	Log       telegraf.Logger
+
+	client       *Client
+	resourceTags []*telegraf.Tag
+	grpcTimeout  time.Duration
+}
+
+const (
+	// QuotaLabelsPerMetricDescriptor is the limit
+	// to labels (tags) per metric descriptor.
+	QuotaLabelsPerMetricDescriptor = 10
+	// QuotaStringLengthForLabelKey is the limit
+	// to string length for label key.
+	QuotaStringLengthForLabelKey = 100
+	// QuotaStringLengthForLabelValue is the limit
+	// to string length for label value.
+	QuotaStringLengthForLabelValue = 1024
+
+	// StartTime for cumulative metrics.
+	StartTime = int64(1)
+	// MaxInt is the max int64 value.
+	MaxInt = int(^uint(0) >> 1)
+
+	defaultEndpoint            = "http://localhost:4317"
+	defaultTimeout             = time.Second * 60
+	instrumentationLibraryName = "Telegraf"
+)
+
+var sampleConfig = `
+  ## OpenTelemetry endpoint
+  endpoint = "http://localhost:4317"
+
+  ## Timeout used when sending data over grpc
+  timeout = "10s"
+
+  # Additional resource attributes
+  [outputs.otlp.attributes]
+  	"service.name" = "demo"
+
+  # Additional grpc metadata
+  [outputs.otlp.headers]
+    key1 = "value1"
+
+`
+
+// Connect initiates the primary connection to the OTLP endpoint.
+func (o *OTLP) Connect() error {
+	if o.Endpoint == "" {
+		o.Endpoint = defaultEndpoint
+	}
+	endpoint, err := url.Parse(o.Endpoint)
+	if err != nil {
+		return fmt.Errorf("invalid endpoint configured")
+	}
+
+	if o.Timeout == "" {
+		o.grpcTimeout = defaultTimeout
+	} else {
+		o.grpcTimeout, err = time.ParseDuration(o.Timeout)
+		if err != nil {
+			return fmt.Errorf("invalid timeout configured")
+		}
+	}
+
+	for k, v := range o.Attributes {
+		o.resourceTags = append(o.resourceTags, &telegraf.Tag{Key: k, Value: v})
+	}
+
+	if o.Headers == nil {
+		o.Headers = make(map[string]string, 1)
+	}
+
+	o.Headers["telemetry-reporting-agent"] = fmt.Sprint(
+		"telegraf/",
+		internal.Version(),
+	)
+
+	if o.client == nil {
+		ctx := context.Background()
+		o.client = NewClient(ClientConfig{
+			URL:     endpoint,
+			Headers: metadata.New(o.Headers),
+			Timeout: o.grpcTimeout,
+		})
+		if err := o.client.Selftest(ctx); err != nil {
+			_ = o.client.Close()
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Sorted returns a copy of the metrics in time ascending order.  A copy is
+// made to avoid modifying the input metric slice since doing so is not
+// allowed.
+func sorted(metrics []telegraf.Metric) []telegraf.Metric {
+	batch := make([]telegraf.Metric, 0, len(metrics))
+	for i := len(metrics) - 1; i >= 0; i-- {
+		batch = append(batch, metrics[i])
+	}
+	sort.Slice(batch, func(i, j int) bool {
+		return batch[i].Time().Before(batch[j].Time())
+	})
+	return batch
+}
```

Comment:
This might be an interesting function to have in the accumulator! Anyway I think you should rather return a sorted set of pointers instead of the full metric copies.

Reply:
This was code leftover from the initial copy of the stackdriver output plugin. Ordering doesn't matter for sending data over OTLP.

Is it worth opening an issue to track the change to a slice of pointers? https://github.com/influxdata/telegraf/blob/ad942052f06bb13ce8dad899cba9066e9ad307ed/plugins/outputs/stackdriver/stackdriver.go#L106-L115